### PR TITLE
cmake-variants functionality

### PR DESCRIPTION
**Pull request by mark2185**
Just an initial commit so it's up here, it's still work in progress.

### CODE REVIEW

**Comment by mark2185** [comment]
I would also be able to reference this custom `build_type` name in creating the `build` build folder using the plugin.

E.g. `/home/user/workspace/build/cmake4vim/CustomBuildTypeName/`

**Comment by mark2185** [comment]
@ilyachur please take a look and let me know what you think!

File: README.md [hunk 1]
```diff
@@ -81,6 +81,7 @@ Plugin supports special global variables which are allow to change behaviour of
  - **`g:cmake_compile_commands_link`** set the path for a link on compile_commands.json. Default is empty.
  - **`g:cmake_build_executor`** allows to force set the build executor. Available values are 'job', 'dispatch', 'system' and ''. Default is empty.
  - **`g:cmake_vimspector_support`** enables generation and modification of [Vimspector](https://github.com/puremourning/vimspector) config file. Default is 0. **Attention! The support of Vimspector config is an experimental feature.**
+ - **`g:cmake_variants`** enables predefined cmake build variants in the form of a dictionary, e.g. `{ 'Debug' : { 'cmake_build_type' : 'Debug', 'cmake_usr_args' : '-DCONAN_PATH=~/.conan' } }`
```
Comment (ilyachur) [inline README.md:84]: I think we cannot change the cmake_build_type for default targets. Maybe we need other example.

File: autoload/cmake4vim.vim [hunk 2]
```diff
@@ -32,10 +32,27 @@ function! cmake4vim#CompleteTarget(arg_lead, cmd_line, cursor_pos) abort
     return join(l:sorted_targets, "\n")
 endfunction
 
+function! cmake4vim#populateDefaultCMakeVariants() abort
```
Comment (ilyachur) [inline autoload/cmake4vim.vim:35]: I think it can be a private function `s:functionName()`

File: autoload/cmake4vim.vim [hunk 3]
```diff
@@ -32,10 +32,27 @@ function! cmake4vim#CompleteTarget(arg_lead, cmd_line, cursor_pos) abort
...
+    for build_type in filter( utils#cmake#getDefaultBuildTypes(), "v:val !=# ''" )
```
Comment (ilyachur) [inline autoload/cmake4vim.vim:36]: Am I right that you try to remove all empty targets? Is it a real case?

File: autoload/cmake4vim.vim [hunk 4]
```diff
@@ -32,10 +32,27 @@ function! cmake4vim#CompleteTarget(arg_lead, cmd_line, cursor_pos) abort
...
+        if ( !has_key( g:cmake_variants, build_type ) )
+            let g:cmake_variants[ build_type ] =
+                \ {
+                \   'cmake_build_type' : build_type,
```
Comment (ilyachur) [inline autoload/cmake4vim.vim:40]: What if user in own config has changed build type for default target. I think we should skip these changes.

File: autoload/cmake4vim.vim [hunk 5]
```diff
@@ -176,7 +193,13 @@ endfunction
 
 " Functions allows to switch between build types
 function! cmake4vim#SelectBuildType(buildType) abort
-    let g:cmake_build_type = a:buildType
+    " so it's called only once
+    if ( !exists( 's:initialized_default_build_types' ) )
+        call cmake4vim#populateDefaultCMakeVariants()
```
Comment (ilyachur) [inline autoload/cmake4vim.vim:198]: I think we can call this method every time, because user can change cmake build types configuration in the runtime.

File: autoload/cmake4vim.vim [hunk 6]
```diff
@@ -176,7 +193,13 @@ endfunction
...
+    endif
+    let s:initialized_default_build_types = v:true
+    let g:cmake_build_type = g:cmake_variants[ a:buildType ][ 'cmake_build_type' ]
```
Comment (ilyachur) [inline autoload/cmake4vim.vim:201]: Maybe we can add private function `s:getCMakeVariants()` which should return up-to-date dictionary.

File: plugin/cmake4vim.vim [hunk 7]
```diff
@@ -63,6 +63,8 @@ if !exists('g:cmake_vimspector_support')
     let g:cmake_vimspector_support = 0
 endif
 
+let g:cmake_variants = get( g:, 'cmake_variants', {} )
```
Comment (ilyachur) [inline plugin/cmake4vim.vim:66]: Hmm, I think we can use this approach for all options. We can do it in the separate PR.

File: autoload/cmake4vim.vim (hunk 3 above)
Comment (mark2185) [reply autoload/cmake4vim.vim:36]: `utils#cmake#getDefaultBuildTypes()` returns `[ 'Release', 'Debug', 'RelWithDebInfo', 'MinSizeRel', '' ]`, I just wanted to discard the last, empty, build type.

**Comment by ilyachur** [comment]
> @ilyachur please take a look and let me know what you think!
> 
> Also, I have no idea why tests are failing, _again_

As I understand in tests we get incorrect number of build targets.
Did you run tests locally?

File: README.md (hunk 1 above)
Comment (mark2185) [reply README.md:84]: I wanted to enable to user to override default build types and add custom ones.
```
{
    'Debug' : { 'cmake_build_type' : 'Debug', 'cmake_usr_args' : '-blahblah' }
}
```
... would override the default `Debug` build, but `Release, RelWithDebInfo, MinSizeRel` still have the default values specified by `g:cmake_usr_args` and `g:cmake_build_type`.

Because if someone wants to have custom `debug`, `release`, `minSizeRel`, why not allow them to override the existing ones, instead of having 3 they will never use which are similarly named as the 3 custom ones.

**Comment by mark2185** [comment]
> As I understand in tests we get incorrect number of build targets.
> Did you run tests locally?

I did, but I didn't get the chance to investigate which tests are failing, and for some reason I cannot run them locally on `macOS` so I'll have to check again when I get to my linux laptop.

**Comment by ilyachur** [comment]
> > As I understand in tests we get incorrect number of build targets.
> > Did you run tests locally?
> 
> I did, but I didn't get the chance to investigate which tests are failing, and for some reason I cannot run them locally on `macOS` so I'll have to check again when I get to my linux laptop.

Hmm, do you use [local_run.sh](https://github.com/ilyachur/cmake4vim/blob/master/test/local_run.sh) script to local run? Because I use it to run all tests on macOS

**Comment by mark2185** [comment]
> Hm, do you use [local_run.sh](https://github.com/ilyachur/cmake4vim/blob/master/test/local_run.sh) script to local run? Because I use it to run all tests on macOS

Ah, I was missing `realpath` command. I installed it now and will get back to you when I figure out what is wrong.

**Comment by mark2185** [comment]
Uhhh, @ilyachur do you have a windows machine available to see which test takes too long?

**Comment by ilyachur** [comment]
> Uhhh, @ilyachur do you have a windows machine available to see which test takes too long?

Usually it is not reproduced on my local machine, as I see from the neovim log the last test was: `    ( 75/100) [EXECUTE] CTest with default settings`.
But it strange because usually (if I implement new feature) I faced with such problems for new tests. And usually it means that something wrong with new functionality or with new tests.
You can try to disable all new tests for windows and after that try to enable them.

**Comment by mark2185** [comment]
Alright, I'll try to disable them, but `CTest with default settings` is not a new test.

**Comment by mark2185** [comment]
Just a reminder to myself to implement `fzf#cmake4vim#SelectBuildType`.

**Comment by mark2185** [comment]
Also this:
> I would also be able to reference this custom build_type name in creating the build build folder using the plugin.

The current `CMakeBuild` still generates `Debug` or `Release` folders, it doesn't take into consideration the custom names.

**Comment by mark2185** [comment]
@ilyachur could the CI be set up in such a way it doesn't cancel all other jobs if one fails? 

In this case, `cmake` failed to install on `macOS` and took all other jobs down with it.

**Comment by mark2185** [comment]
> The current `CMakeBuild` still generates `Debug` or `Release` folders, it doesn't take into consideration the custom names.

@ilyachur could we introduce a new variable, e.g. `build_name` which would be set to the selected `cmake_variants` key and change `s:detectCMakeBuildDir()` to use `g:build_name` instead of `l:build_type`?

It would still work the same for the default build types, but users could create custom ones in `cmake_variants` and then the build folders would have matching names?

**Comment by ilyachur** [comment]
> Also this:
> 
> > I would also be able to reference this custom build_type name in creating the build build folder using the plugin.
> 
> The current `CMakeBuild` still generates `Debug` or `Release` folders, it doesn't take into consideration the custom names.

I think we can use the name of build in order to generate build folder. Because the set of names will be bigger than set of build types and moreover we always have default build types in the namespace of build names.

**Comment by ilyachur** [comment]
> @ilyachur could the CI be set up in such a way it doesn't cancel all other jobs if one fails?
> 
> In this case, `cmake` failed to install on `macOS` and took all other jobs down with it.

I don't know. I think we can try to add this flag: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idcontinue-on-error

**Comment by ilyachur** [comment]
> @ilyachur could we introduce a new variable, e.g. `build_name` which would be set to the selected `cmake_variants` key and change `s:detectCMakeBuildDir()` to use `g:build_name` instead of `l:build_type`?
> 
> It would still work the same for the default build types, but users could create custom ones in `cmake_variants` and then the build folders would have matching names?

Can we use for this purpose cmake_build_type variable? Because in this case we can avoid the introducing of new variables and from the plugin perspective (I think) custom build types are the same types as default build types.

**Comment by mark2185** [comment]
> Can we use for this purpose cmake_build_type variable? Because in this case we can avoid the introducing of new variables and from the plugin perspective (I think) custom build types are the same types as default build types.

How exactly should `s:detectCMakeBuildType` work then? It can only see the actual build type, but for getting the name back, we would have to parse the path to get the "name" out.

Or disable detection of build type from existing build folders, but that sounds useful, I don't want to remove that.

**Comment by ilyachur** [comment]
> > Can we use for this purpose cmake_build_type variable? Because in this case we can avoid the introducing of new variables and from the plugin perspective (I think) custom build types are the same types as default build types.
> 
> How exactly should `s:detectCMakeBuildType` work then? It can only see the actual build type, but for getting the name back, we would have to parse the path to get the "name" out.
> 
> Or disable detection of build type from existing build folders, but that sounds useful, I don't want to remove that.

I don't see any issues with `s:detectCMakeBuildType` function. I think we can expand it. At the current moment we use hardcoded names in this method to detect build folders, I think we can extend this method and also use custom names in order to find build folder.

**Comment by mark2185** [comment]
Found a bug, if you set the `g:cmake_build_type` variable to e.g. `debug` in your `.vimrc`, and run `:CMake`, the `cmake_variants` dictionary isn't used.

**Comment by mark2185** [comment]
@ilyachur could you please take a look at this? I don't see which assert failed [on the windows build](https://github.com/ilyachur/cmake4vim/pull/80/checks?check_run_id=2191701584), and I don't have a windows machine where I can debug that.

I'm only asking because I have another PR ready, one for [cmake-kits](https://vector-of-bool.github.io/docs/vscode-cmake-tools/kits.html) like we have discussed before, and I would like to rebase it onto this.

**Comment by mark2185** [comment]
Moved to https://github.com/ilyachur/cmake4vim/pull/81 .
